### PR TITLE
Add support for tainting headers

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Add support for tainting of header names and values

# Motivation

# Additional Notes

### CODE REVIEW

**Comment by jandro996** [comment]
Seems that we can reuse the same code for Request hedaders, parameters and cookies as we will  taint name and value

File: dd-java-agent/instrumentation/servlet/src/test/groovy/HttpServletRequestCallSiteTest.groovy [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import datadog.trace.agent.test.AgentTestRunner
+import datadog.trace.api.Platform
+import datadog.trace.api.iast.InstrumentationBridge
+import datadog.trace.api.iast.source.WebModule
+import spock.lang.IgnoreIf
+
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletRequestWrapper
+
+@IgnoreIf({
```
Comment (jandro996) [inline dd-java-agent/instrumentation/servlet/src/test/groovy/HttpServletRequestCallSiteTest.groovy:11]: I think we  can remove this kind of validations for < jdk8

**Review by jandro996 — APPROVED** [review]

File: dd-java-agent/instrumentation/servlet/src/test/groovy/HttpServletRequestCallSiteTest.groovy (hunk 1 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/servlet/src/test/groovy/HttpServletRequestCallSiteTest.groovy:11]: Yep, https://github.com/DataDog/dd-trace-java/pull/4418

let me change it

File: dd-java-agent/instrumentation/servlet/src/main/java/datadog/trace/instrumentation/servlet/request/HttpServletRequestCallSite.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    if (enumeration == null) {
+      return null;
+    }
+    final WebModule module = InstrumentationBridge.WEB;
+    if (module == null) {
+      return enumeration;
+    }
+    final List<String> headerNames = new ArrayList<>();
+    while (enumeration.hasMoreElements()) {
+      final String headerName = (String) enumeration.nextElement();
```
Comment (smola) [inline dd-java-agent/instrumentation/servlet/src/main/java/datadog/trace/instrumentation/servlet/request/HttpServletRequestCallSite.java:84]: This should probably be within the try/catch block?

Comment (smola) [reply dd-java-agent/instrumentation/servlet/src/main/java/datadog/trace/instrumentation/servlet/request/HttpServletRequestCallSite.java:84]: Maybe also the `hasMoreElements()`...

Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/servlet/src/main/java/datadog/trace/instrumentation/servlet/request/HttpServletRequestCallSite.java]: Nice catch, it's a bit more complex like move it withing the try/catch block. If they fail it's servlet code that fails and we should let the exception propagate removing ourselves from the stack first. Check the latest version and the specific tests for that matter:

```java
  def 'test that get headers does not fail when servlet related code fails'(final Class<? extends HttpServletRequest> clazz) {
    setup:
    final iastModule = Mock(WebModule)
    InstrumentationBridge.registerIastModule(iastModule)
    final enumeration = Mock(Enumeration) {
      hasMoreElements() >> { throw new NuclearBomb('Boom!!!')}
    }
    final testSuite = new TestHttpServletRequestCallSiteSuite(Mock(clazz) {
      getHeaders('header') >> enumeration
    })

    when:
    testSuite.getHeaders('header')

    then:
    final bomb = thrown(NuclearBomb)
    bomb.stackTrace.find { it.className == HttpServletRequestCallSite.name } == null

    where:
    clazz                     | _
    HttpServletRequest        | _
    HttpServletRequestWrapper | _
  }
```

**Review by smola — APPROVED** [review]

**Pull request merged**